source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -134,15 +135,27 @@ Http::FilterHeadersStatus CacheFilter::encodeHeaders(Http::ResponseHeaderMap& he
   if (request_allows_inserts_ && !is_head_request_ &&
       CacheabilityUtils::isCacheableResponse(headers, vary_allow_list_)) {
     ENVOY_STREAM_LOG(debug, "CacheFilter::encodeHeaders inserting headers", *encoder_callbacks_);
-    insert_ = cache_->makeInsertContext(std::move(lookup_), *encoder_callbacks_);
-    // Add metadata associated with the cached response. Right now this is only response_time;
-    const ResponseMetadata metadata = {time_source_.systemTime()};
-    // TODO(capoferro): Note that there is currently no way to communicate back to the CacheFilter
-    // that an insert has failed. If an insert fails partway, it's better not to send additional
-    // chunks to the cache if we're already in a failure state and should abort, but we can only do
-    // that if we can communicate failures back to the filter, so we should fix this.
-    insert_->insertHeaders(
-        headers, metadata, [](bool) {}, end_stream);
+    auto insert_context = cache_->makeInsertContext(std::move(lookup_), *encoder_callbacks_);
+    if (insert_context) {
+      // The callbacks passed to CacheInsertQueue are all called through the dispatcher,
+      // so they're thread-safe, and onDestroy will prevent them from being called via
+      // giving the queue ownership of itself, so they're also filter-destruction-safe.
+      insert_queue_ = std::make_unique<CacheInsertQueue>(
+          decoder_callbacks_->dispatcher(), std::move(insert_context),
+          encoder_callbacks_->encoderBufferLimit(), encoder_callbacks_->encoderBufferLimit() / 2,
+          // High watermark callback.
+          [this]() { encoder_callbacks_->onEncoderFilterAboveWriteBufferHighWatermark(); },
```

Comment:
is there some reason we are indirecting through a function pointer here, as opposed to just passing encoder_callbacks_ into CacheInsertQueue and having it call those directly? I think that would drop 4 args from CacheInsertQueue.

I'm sure there is one, but could you comment on that, either here or in the definition of CacheInsertQueue?